### PR TITLE
CRYP-68: implement eth edge api for adding wallets

### PR DESCRIPTION
**Pull request by Andre-Ibrahim**

### CODE REVIEW

File: packages/api/src/wallets/guards/authorization.guard.ts [hunk 1]
```diff
@@ -4,8 +4,10 @@ import { Observable } from "rxjs";
 @Injectable()
 export class AuthorizationGuard implements CanActivate {
     canActivate(context: ExecutionContext): boolean | Promise<boolean> | Observable<boolean> {
-        const { params, user } = context.switchToHttp().getRequest();
-        if (params.id != user.id) throw new UnauthorizedException();
+        const { body, params, user } = context.switchToHttp().getRequest();
+        if (params.id != user.id || params.id != body.userId) {
```
Comment (jason-gerard) [inline packages/api/src/wallets/guards/authorization.guard.ts:8]: Maybe add a comment here explaining that this check makes sure the user id in the token, url params, and body are all the same

File: packages/api/src/gateways/eth_edge_gateway.ts [hunk 2]
```diff
@@ -12,6 +12,6 @@ export class EthEdgeGateway {
 
     async createWallet<T>(req: CreateWalletRequest): Promise<T> {
         const path = `user/${req.userId}/wallet`;
-        return this.request<any>(Method.POST, path, req);
+        return this.request<T>(Method.POST, path, req);
```
Comment (jason-gerard) [inline packages/api/src/gateways/eth_edge_gateway.ts:15]: This method doesn't need to be generic, it always returns the same type which is `WalletWithBalance`

File: packages/api/src/wallets/wallets.service.ts [hunk 3]
```diff
@@ -15,10 +16,10 @@ export class WalletsService {
         private readonly ethEdgeGateway: EthEdgeGateway,
     ) {}
 
-    async create(req: CreateWalletRequest): Promise<InsertResult> {
+    async create(req: CreateWalletRequest): Promise<WalletWithBalance> {
         if (req.currencyType != getCurrencyType(req.address))
```
Comment (jason-gerard) [inline packages/api/src/wallets/wallets.service.ts:20]: Curly braces? Style guide?

File: packages/eth-edge/src/modules/app.module.ts [hunk 4]
```diff
@@ -13,10 +13,13 @@ import { WalletsModule } from "./wallets/wallets.module";
         }),
         TypeOrmModule.forRootAsync({
             inject: [ConfigService],
-            useFactory: (config: ConfigService) => ({
-                ...dataSourceOptionsConfig(config),
-                synchronize: true,
-            }),
+            useFactory: (config: ConfigService) => {
+                console.log(config);
```
Comment (jason-gerard) [inline packages/eth-edge/src/modules/app.module.ts:17]: Extra console.log

```suggestion
```

File: packages/common/src/errors/error_messages.ts [hunk 5]
```diff
@@ -2,3 +2,4 @@ export const ERROR_EMAIL_OR_PASSWORD_INCORRECT = "Your email or password was inc
 export const ERROR_EMAIL_IN_USE = "This email address is already associated with another account.";
 export const ERROR_NOP = "nop";
 export const ERROR_WALLET_ADDRESS_FOR_CURRENCY = "Wallet address is not valid for currency type.";
+export const ERROR_WALLET_ALREADY_ADDED_TO_ACCOUNT = "Wallet was already added to this account.";
```
Comment (domsec) [inline packages/common/src/errors/error_messages.ts:5]: Great catch on this potential error as I forgot about it when writing the specs. Can you please change it to "This wallet address is already associated with another wallet.". Specs will be updated accordingly.

File: packages/eth-edge/src/services/wallets.service.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+    async create(createWalletReq: CreateWalletRequest): Promise<WalletWithBalance> {
+        if (await this.findOne(createWalletReq.address, createWalletReq.userId)) {
+            throw new BadRequestException(ERROR_WALLET_ALREADY_ADDED_TO_ACCOUNT);
+        }
+
+        const wallet = this.walletRepository.create(createWalletReq);
+        await this.walletRepository.insert(wallet);
+        const balance = await this.alchemyNodeService.getBalance(createWalletReq.address);
+        return { ...(await this.findOne(createWalletReq.address, createWalletReq.userId)), balance };
```
Comment (jason-gerard) [inline packages/eth-edge/src/services/wallets.service.ts:26]: Just a little bit of reformatting to make the code easier to read

```suggestion
        const reqWallet = this.walletRepository.create(createWalletReq);
        await this.walletRepository.insert(wallet);
        
        const balance = await this.alchemyNodeService.getBalance(createWalletReq.address);
        const wallet = await this.findOne(createWalletReq.address, createWalletReq.userId);
        return {
            ...wallet,
            balance
        };
```

File: packages/common/src/errors/error_messages.ts [hunk 7]
```diff
@@ -2,3 +2,4 @@ export const ERROR_EMAIL_OR_PASSWORD_INCORRECT = "Your email or password was inc
 export const ERROR_EMAIL_IN_USE = "This email address is already associated with another account.";
 export const ERROR_NOP = "nop";
 export const ERROR_WALLET_ADDRESS_FOR_CURRENCY = "Wallet address is not valid for currency type.";
```
Comment (domsec) [inline packages/common/src/errors/error_messages.ts:4]: As I just noticed this error message in this PR, even though it's not related can it please be changed to "Enter a valid CURRENCY_TYPE wallet address." (i.e. "Enter a valid Ethereum wallet address." Specs will be updated accordingly.

**Review by domsec — APPROVED** [review]

**Pull request merged**